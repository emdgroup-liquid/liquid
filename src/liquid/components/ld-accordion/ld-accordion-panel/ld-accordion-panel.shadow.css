:host {
  /* Not using 0s here because we rely on the transition end event. */
  --ld-accordion-panel-transition-duration: var(
    --ld-transition-duration-instant
  );

  position: relative;
  display: block;
  box-sizing: border-box;
  max-height: var(--ld-accordion-panel-max-height, auto);
  overflow: hidden;
  background-color: var(--ld-accordion-bg-col);
  border-radius: var(--ld-accordion-border-radius);
  border-top-left-radius: inherit;
  border-top-right-radius: inherit;
  will-change: max-height;

  &(.ld-accordion-panel--initialized) {
    /* stylelint-disable-next-line plugin/no-low-performance-animation-properties */
    transition: max-height var(--ld-accordion-panel-transition-duration) ease;

    @media (prefers-reduced-motion: no-preference) {
      --ld-accordion-panel-transition-duration: var(
        --ld-transition-duration-normal
      );
    }
  }

  &(.ld-accordion-panel--expanded) {
    &::before {
      content: '';
      height: var(--ld-sp-2);
      background-color: var(--ld-accordion-panel-border-top-col);
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
    }

    .ld-accordion-panel__content {
      opacity: 1;
      visibility: inherit;
      transition: opacity var(--ld-accordion-panel-transition-duration) linear;
    }
  }
}

.ld-accordion-panel__content {
  opacity: 0;
  visibility: hidden;
  transition: opacity var(--ld-accordion-panel-transition-duration) linear,
    visibility 0s var(--ld-accordion-panel-transition-duration) linear;
}
