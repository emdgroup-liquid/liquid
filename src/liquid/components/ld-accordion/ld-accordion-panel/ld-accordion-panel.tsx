import '../../../components' // type definitions for type checks and intelliSense
import { Component, Element, h, Host, Method, State } from '@stencil/core'
import { getClassNames } from '../../../utils/getClassNames'
import { closest } from '../../../utils/closest'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: 'ld-accordion-panel',
  styleUrl: 'ld-accordion-panel.shadow.css',
  shadow: true,
})
export class LdAccordionPanel {
  @Element() el: HTMLElement

  // Container to be observed for size changes.
  // Note that we can not observe size changes on the element itself,
  // As with the max-height prop applied to it changes to the content
  // will not trigger a resize event. Hence, we use a container element.
  private contentRef: HTMLDivElement

  @State() expanded: boolean
  @State() initialized = false
  @State() transitionEnabled = false
  @State() maxHeight: number
  @State() resizeObserver: ResizeObserver
  @State() innerPanelExpanding = false

  /**
   * @internal
   * Updates expanded state.
   */
  @Method()
  async applyMaxHeight(additionalHeightFromInnerPanel = 0) {
    if (additionalHeightFromInnerPanel) {
      this.innerPanelExpanding = true
    }

    // Apply max height on outer panel inside nested accordion.
    if (this.expanded) {
      const closestPanel = closest('ld-accordion-panel', this.el.parentElement)
      closestPanel?.applyMaxHeight(this.el.scrollHeight)
    }

    this.maxHeight = this.expanded
      ? this.el.scrollHeight + additionalHeightFromInnerPanel
      : 0
  }

  /**
   * @internal
   * Updates expanded state.
   */
  @Method()
  async setExpanded(expanded: boolean) {
    this.expanded = expanded

    this.applyMaxHeight()
  }

  private onTransitionEnd = (ev: TransitionEvent) => {
    if (ev.target === this.el) {
      this.innerPanelExpanding = false
    }
  }

  componentDidLoad() {
    setTimeout(() => {
      this.resizeObserver = new ResizeObserver(() => {
        // When a panel is expanding inside a nested accordion, the nested panel
        // takes over the responsibility for updating the max-height on the outer
        // panel. In other words: We disable the observer callback function in
        // order to instantly update the max-height for a better performance.
        if (!this.innerPanelExpanding) {
          this.applyMaxHeight()
        }
      })
      this.resizeObserver.observe(this.contentRef)
      this.initialized = true
    })
  }

  disconnectedCallback() {
    this.resizeObserver.unobserve(this.contentRef)
  }

  render() {
    const cl = getClassNames([
      'ld-accordion-panel',
      this.expanded && 'ld-accordion-panel--expanded',
      this.initialized && 'ld-accordion-panel--initialized',
    ])

    return (
      <Host
        style={{ '--ld-accordion-panel-max-height': this.maxHeight + 'px' }}
        class={cl}
        onTransitionEnd={this.onTransitionEnd}
      >
        <div
          ref={(ref) => (this.contentRef = ref)}
          class="ld-accordion-panel__content"
        >
          <slot></slot>
        </div>
      </Host>
    )
  }
}
