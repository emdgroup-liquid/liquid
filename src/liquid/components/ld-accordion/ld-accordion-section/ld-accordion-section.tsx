import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Prop,
  State,
  Watch,
} from '@stencil/core'
import { getClassNames } from '../../../utils/getClassNames'
import { closest } from '../../../utils/closest'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: 'ld-accordion-section',
  styleUrl: 'ld-accordion-section.shadow.css',
  shadow: true,
})
export class LdAccordionSection {
  @Element() el: HTMLElement

  /** Indicates that the accordion section is expanded. */
  @Prop({ mutable: true }) expanded?: boolean

  @State() initialized = false

  /** Emitted on expansion and collapse. */
  @Event() ldaccordionchange: EventEmitter<boolean>

  @Watch('expanded')
  updateExpandedState(newExpanded: boolean) {
    Array.from(this.el.children).forEach(
      (child: HTMLLdAccordionPanelElement | HTMLLdAccordionToggleElement) => {
        if (typeof child.setExpanded === 'function') {
          child.setExpanded(newExpanded)
        }
      }
    )

    if (this.initialized) {
      this.ldaccordionchange.emit(newExpanded)
    }
  }

  private handleToggleClick(ev) {
    // closest utility function must be used here for the component
    // to work in Solid.js app, where ev.target can be an element
    // within the shadow DOM of the component.
    // Usage of ev.composedPath() is required for penetrating shadow DOM.
    const target = 'composedPath' in ev ? ev.composedPath().at(0) : ev.target
    if (closest('ld-accordion-section', target) !== this.el) {
      return
    }
    this.expanded = !this.expanded
  }

  componentWillLoad() {
    this.updateExpandedState(this.expanded)
    this.initialized = true
  }

  render() {
    const cl = getClassNames([
      'ld-accordion-section',
      this.expanded && 'ld-accordion-section--expanded',
    ])

    return (
      <Host
        class={cl}
        onLdaccordiontoggleclick={this.handleToggleClick.bind(this)}
      >
        <slot></slot>
      </Host>
    )
  }
}
