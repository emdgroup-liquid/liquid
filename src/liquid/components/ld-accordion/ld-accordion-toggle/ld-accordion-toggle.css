:host {
  background-color: var(--ld-col-wht);
  display: block;
  border-radius: var(--ld-accordion-border-radius);
}

.ld-accordion-toggle__button,
.ld-accordion-toggle__label,
.ld-accordion-toggle__trigger {
  margin: 0;
  border: 0;
  box-sizing: border-box;
  text-align: left;
  touch-action: manipulation;
  user-select: none;
  width: 100%;
  -webkit-touch-callout: none;
}

.ld-accordion-toggle__button {
  background-color: var(--ld-accordion-bg-col);
  align-items: center;
  border-radius: inherit;
  color: var(--ld-accordion-col);
  display: grid;
  font: var(--ld-typo-body-m);
  grid-template-columns: 1fr var(--ld-accordion-toggle-size);
  padding: 0;
  position: relative;

  &[aria-disabled] {
    color: var(--ld-accordion-col-disabled);
  }

  &:not([aria-disabled]) {
    cursor: pointer;

    &:where(:focus:focus-visible) {
      background-color: var(--ld-accordion-bg-col-focus);
      color: var(--ld-accordion-col-focus);
    }
    @media (hover: hover) {
      &:where(:hover) {
        background-color: var(--ld-accordion-bg-col-hover);
        color: var(--ld-accordion-col-hover);
      }
    }
    &:where(:active),
    &:where(:active:focus-visible) {
      background-color: var(--ld-accordion-bg-col-active);
      color: var(--ld-accordion-col-active);
    }

    .ld-accordion-toggle__trigger {
      color: var(--ld-accordion-trigger-icon-col);
    }

    &:focus:focus-visible,
    &:hover,
    &:active,
    &:active:focus-visible {
      .ld-accordion-toggle__trigger {
        color: inherit;
      }
    }
  }
}

.ld-accordion-toggle__label,
.ld-accordion-toggle__trigger {
  font: inherit;
  background-color: transparent;

  &:not([aria-disabled]) {
    cursor: pointer;
  }
}

.ld-accordion-toggle__label {
  position: relative;
  padding: var(--ld-accordion-padding-y) var(--ld-accordion-padding-x);
}

.ld-accordion-toggle__trigger {
  position: relative;
  width: var(--ld-accordion-toggle-size);
  height: var(--ld-accordion-toggle-size);
  display: grid;
  place-items: center;
  overflow: hidden; /* resolves focus outline issue on safari */
}

.ld-accordion-toggle__label-content,
.ld-accordion-toggle__trigger-content {
  position: relative;
  display: grid;
}

:host(.ld-accordion-toggle--expanded) {
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;

  .ld-accordion-toggle__trigger-icon {
    transform: scaleY(-1);
  }
}

:host(.ld-accordion-toggle--split) {
  .ld-accordion-toggle__label,
  .ld-accordion-toggle__trigger {
    &::before {
      content: '';
      position: absolute;
      inset: var(--ld-sp-4);
      border-radius: var(--ld-br-s);
    }

    &[aria-disabled] {
      color: var(--ld-accordion-col-disabled);
    }
  }

  .ld-accordion-toggle__button:not([aria-disabled]) {
    background-color: var(--ld-accordion-bg-col);

    .ld-accordion-toggle__label:not([aria-disabled]) {
      color: var(--ld-accordion-col);
    }
    .ld-accordion-toggle__trigger:not([aria-disabled]) {
      color: var(--ld-thm-primary);
    }

    .ld-accordion-toggle__label:not([aria-disabled]),
    .ld-accordion-toggle__trigger:not([aria-disabled]) {
      &:where(:focus:focus-visible) {
        color: var(--ld-accordion-col-focus);
        &::before {
          background-color: var(--ld-accordion-bg-col-focus);
        }
      }
      @media (hover: hover) {
        &:where(:hover) {
          color: var(--ld-accordion-col-hover);
          &::before {
            background-color: var(--ld-accordion-bg-col-hover);
          }
        }
      }
      &:where(:active),
      &:where(:active:focus-visible) {
        color: var(--ld-accordion-col-active);
        &::before {
          background-color: var(--ld-accordion-bg-col-active);
        }
      }
    }
  }
}

.ld-accordion-toggle__content {
  display: contents;
}
