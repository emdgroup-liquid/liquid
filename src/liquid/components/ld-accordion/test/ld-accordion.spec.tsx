import { h } from '@stencil/core'
import { newSpecPage } from '@stencil/core/testing'
import { LdAccordion } from '../ld-accordion'
import { LdAccordionPanel } from '../ld-accordion-panel/ld-accordion-panel'
import { LdAccordionSection } from '../ld-accordion-section/ld-accordion-section'
import { LdAccordionToggle } from '../ld-accordion-toggle/ld-accordion-toggle'
import { getTriggerableResizeObserver } from '../../../utils/resizeObserver'

const components = [
  LdAccordion,
  LdAccordionSection,
  LdAccordionToggle,
  LdAccordionPanel,
]

async function transitionEnd(page, target) {
  const transitionEndHandler = page.root['__listeners'].find(
    (l) => l.type === 'transitionEnd'
  ).handler
  transitionEndHandler({ target })
  await page.waitForChanges()
}

describe('ld-accordion', () => {
  beforeEach(() => {
    jest.clearAllTimers()
  })

  it('renders', async () => {
    const page = await newSpecPage({
      components,
      template: () => (
        <ld-accordion>
          <ld-accordion-section expanded>
            <ld-accordion-toggle>Fruits</ld-accordion-toggle>
            <ld-accordion-panel>Apple, orange, banana</ld-accordion-panel>
          </ld-accordion-section>
          <ld-accordion-section>
            <ld-accordion-toggle>Vegetables</ld-accordion-toggle>
            <ld-accordion-panel>Potato, cucumber, tomato</ld-accordion-panel>
          </ld-accordion-section>
          <ld-accordion-section>
            <ld-accordion-toggle disabled>Nuts</ld-accordion-toggle>
            <ld-accordion-panel>
              Walnut, chestnut, strawberry
            </ld-accordion-panel>
          </ld-accordion-section>
        </ld-accordion>
      ),
    })
    expect(page.root).toMatchSnapshot()
  })

  it('observes panel content for changes to its size', async () => {
    const page = await newSpecPage({
      components,
      template: () => (
        <ld-accordion>
          <ld-accordion-section>
            <ld-accordion-toggle>Fruits</ld-accordion-toggle>
            <ld-accordion-panel>Apple, orange, banana</ld-accordion-panel>
          </ld-accordion-section>
          <ld-accordion-section>
            <ld-accordion-toggle>Vegetables</ld-accordion-toggle>
            <ld-accordion-panel>Potato, cucumber, tomato</ld-accordion-panel>
          </ld-accordion-section>
          <ld-accordion-section>
            <ld-accordion-toggle disabled>Nuts</ld-accordion-toggle>
            <ld-accordion-panel>
              Walnut, chestnut, strawberry
            </ld-accordion-panel>
          </ld-accordion-section>
        </ld-accordion>
      ),
    })

    expect(getTriggerableResizeObserver()).toBeUndefined()

    const ldAccordion = page.root

    const ldAccordionPanels = ldAccordion.querySelectorAll('ld-accordion-panel')
    expect(ldAccordionPanels.length).toEqual(3)

    const firstPanel = ldAccordionPanels[0]

    expect(firstPanel).not.toHaveClass('ld-accordion-panel--initialized')

    jest.advanceTimersByTime(0)
    await page.waitForChanges()
    await transitionEnd(page, firstPanel)
    expect(firstPanel).toHaveClass('ld-accordion-panel--initialized')

    const resizeObserver = getTriggerableResizeObserver()
    expect(resizeObserver).not.toBeUndefined()

    resizeObserver.trigger()
    expect(firstPanel.style.cssText).toEqual(
      '--ld-accordion-panel-max-height: 0px;'
    )
  })

  describe('mouse interactions', () => {
    it('expands sections via click with preselected section', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-accordion>
            <ld-accordion-section expanded>
              <ld-accordion-toggle>Fruits</ld-accordion-toggle>
              <ld-accordion-panel>Apple, orange, banana</ld-accordion-panel>
            </ld-accordion-section>
            <ld-accordion-section>
              <ld-accordion-toggle>Vegetables</ld-accordion-toggle>
              <ld-accordion-panel>Potato, cucumber, tomato</ld-accordion-panel>
            </ld-accordion-section>
            <ld-accordion-section>
              <ld-accordion-toggle disabled>Nuts</ld-accordion-toggle>
              <ld-accordion-panel>
                Walnut, chestnut, strawberry
              </ld-accordion-panel>
            </ld-accordion-section>
          </ld-accordion>
        ),
      })

      const ldAccordion = page.root

      const ldAccordionToggles = ldAccordion.querySelectorAll(
        'ld-accordion-toggle'
      )
      expect(ldAccordionToggles.length).toEqual(3)

      const ldAccordionPanels =
        ldAccordion.querySelectorAll('ld-accordion-panel')
      expect(ldAccordionPanels.length).toEqual(3)

      const triggerBtn0 =
        ldAccordionToggles[0].shadowRoot.querySelector('button')
      const triggerBtn1 =
        ldAccordionToggles[1].shadowRoot.querySelector('button')

      expect(triggerBtn0.getAttribute('aria-expanded')).toEqual('true')
      expect(triggerBtn1.getAttribute('aria-expanded')).toEqual('false')

      triggerBtn1.dispatchEvent(new Event('click'))
      await page.waitForChanges()
      await transitionEnd(page, ldAccordionPanels[1])

      expect(triggerBtn0).toHaveAttribute('aria-expanded')
      expect(triggerBtn1).toHaveAttribute('aria-expanded')
    })

    it('does not expands on click in nested accordion', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-accordion>
            <ld-accordion-section expanded>
              <ld-accordion-toggle id="toggle-outer">Outer</ld-accordion-toggle>
              <ld-accordion-panel>
                <ld-accordion>
                  <ld-accordion-section>
                    <ld-accordion-toggle id="toggle-inner">
                      Inner
                    </ld-accordion-toggle>
                    <ld-accordion-panel>Yolo</ld-accordion-panel>
                  </ld-accordion-section>
                </ld-accordion>
              </ld-accordion-panel>
            </ld-accordion-section>
          </ld-accordion>
        ),
      })

      const ldAccordion = page.root

      const ldAccordionToggleOuter = ldAccordion.querySelector('#toggle-outer')
      const ldAccordionToggleInner = ldAccordion.querySelector('#toggle-inner')

      const triggerBtnOuter =
        ldAccordionToggleOuter.shadowRoot.querySelector('button')
      const triggerBtnInner =
        ldAccordionToggleInner.shadowRoot.querySelector('button')

      expect(triggerBtnOuter.getAttribute('aria-expanded')).toEqual('true')
      expect(triggerBtnInner.getAttribute('aria-expanded')).toEqual('false')

      triggerBtnInner.dispatchEvent(new Event('click'))
      await page.waitForChanges()

      expect(triggerBtnOuter).toHaveAttribute('aria-expanded')
      expect(triggerBtnInner).toHaveAttribute('aria-expanded')
    })

    it('collapses sections in single mode', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-accordion single>
            <ld-accordion-section expanded>
              <ld-accordion-toggle>Fruits</ld-accordion-toggle>
              <ld-accordion-panel>Apple, orange, banana</ld-accordion-panel>
            </ld-accordion-section>
            <ld-accordion-section>
              <ld-accordion-toggle>Vegetables</ld-accordion-toggle>
              <ld-accordion-panel>Potato, cucumber, tomato</ld-accordion-panel>
            </ld-accordion-section>
            <ld-accordion-section>
              <ld-accordion-toggle disabled>Nuts</ld-accordion-toggle>
              <ld-accordion-panel>
                Walnut, chestnut, strawberry
              </ld-accordion-panel>
            </ld-accordion-section>
          </ld-accordion>
        ),
      })

      const ldAccordion = page.root

      const ldAccordionToggles = ldAccordion.querySelectorAll(
        'ld-accordion-toggle'
      )
      expect(ldAccordionToggles.length).toEqual(3)

      const ldAccordionPanels =
        ldAccordion.querySelectorAll('ld-accordion-panel')
      expect(ldAccordionPanels.length).toEqual(3)

      const triggerBtn0 =
        ldAccordionToggles[0].shadowRoot.querySelector('button')
      const triggerBtn1 =
        ldAccordionToggles[1].shadowRoot.querySelector('button')

      expect(triggerBtn0.getAttribute('aria-expanded')).toEqual('true')
      expect(triggerBtn1.getAttribute('aria-expanded')).toEqual('false')

      triggerBtn1.dispatchEvent(new Event('click'))
      await page.waitForChanges()

      expect(triggerBtn0.getAttribute('aria-expanded')).toEqual('false')
      expect(triggerBtn1.getAttribute('aria-expanded')).toEqual('true')
    })
  })

  describe('keyboard interactions', () => {
    it('changes focus via arrow keys', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-accordion>
            <ld-accordion-section expanded>
              <ld-accordion-toggle>Fruits</ld-accordion-toggle>
              <ld-accordion-panel>Apple, orange, banana</ld-accordion-panel>
            </ld-accordion-section>
            <ld-accordion-section>
              <ld-accordion-toggle>Vegetables</ld-accordion-toggle>
              <ld-accordion-panel>Potato, cucumber, tomato</ld-accordion-panel>
            </ld-accordion-section>
            <ld-accordion-section>
              <ld-accordion-toggle disabled>Nuts</ld-accordion-toggle>
              <ld-accordion-panel>
                Walnut, chestnut, strawberry
              </ld-accordion-panel>
            </ld-accordion-section>
          </ld-accordion>
        ),
      })
      const ldAccordion = page.root

      const ldAccordionToggles = ldAccordion.querySelectorAll(
        'ld-accordion-toggle'
      )
      const [ldAccordionToggle0, ldAccordionToggle1, ldAccordionToggle2] =
        Array.from(ldAccordionToggles)
      expect(ldAccordionToggles.length).toEqual(3)

      const triggerBtn0 = ldAccordionToggle0.shadowRoot.querySelector('button')
      const triggerBtn1 = ldAccordionToggle1.shadowRoot.querySelector('button')
      const triggerBtn2 = ldAccordionToggle2.shadowRoot.querySelector('button')

      triggerBtn0.focus = jest.fn()
      triggerBtn1.focus = jest.fn()
      triggerBtn2.focus = jest.fn()

      ldAccordionToggle0.dispatchEvent(
        new KeyboardEvent('keydown', {
          key: 'ArrowDown',
          bubbles: true,
        })
      )
      await page.waitForChanges()

      expect(triggerBtn0.focus).toHaveBeenCalledTimes(0)
      expect(triggerBtn1.focus).toHaveBeenCalledTimes(1)
      expect(triggerBtn2.focus).toHaveBeenCalledTimes(0)

      ldAccordion.dispatchEvent(
        new KeyboardEvent('keydown', {
          key: 'ArrowDown',
          bubbles: true,
        })
      )
      await page.waitForChanges()

      expect(triggerBtn0.focus).toHaveBeenCalledTimes(0)
      expect(triggerBtn1.focus).toHaveBeenCalledTimes(1)
      expect(triggerBtn2.focus).toHaveBeenCalledTimes(0)

      ldAccordionToggle1.dispatchEvent(
        new KeyboardEvent('keydown', {
          key: 'ArrowDown',
          bubbles: true,
        })
      )
      await page.waitForChanges()

      expect(triggerBtn0.focus).toHaveBeenCalledTimes(0)
      expect(triggerBtn1.focus).toHaveBeenCalledTimes(1)
      expect(triggerBtn2.focus).toHaveBeenCalledTimes(1)

      ldAccordionToggle2.dispatchEvent(
        new KeyboardEvent('keydown', {
          key: 'ArrowDown',
          bubbles: true,
        })
      )
      await page.waitForChanges()

      expect(triggerBtn0.focus).toHaveBeenCalledTimes(0)
      expect(triggerBtn1.focus).toHaveBeenCalledTimes(1)
      expect(triggerBtn2.focus).toHaveBeenCalledTimes(1)

      ldAccordionToggle1.dispatchEvent(
        new KeyboardEvent('keydown', {
          key: 'ArrowUp',
          bubbles: true,
        })
      )
      await page.waitForChanges()

      expect(triggerBtn0.focus).toHaveBeenCalledTimes(1)
      expect(triggerBtn1.focus).toHaveBeenCalledTimes(1)
      expect(triggerBtn2.focus).toHaveBeenCalledTimes(1)

      ldAccordionToggle0.dispatchEvent(
        new KeyboardEvent('keydown', {
          key: 'ArrowUp',
          bubbles: true,
        })
      )
      await page.waitForChanges()

      expect(triggerBtn0.focus).toHaveBeenCalledTimes(1)
      expect(triggerBtn1.focus).toHaveBeenCalledTimes(1)
      expect(triggerBtn2.focus).toHaveBeenCalledTimes(1)
    })
  })

  describe('events', () => {
    it('emits ldaccordionchange and ldaccordiontoggleclick events', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-accordion>
            <ld-accordion-section expanded>
              <ld-accordion-toggle>Fruits</ld-accordion-toggle>
              <ld-accordion-panel>Apple, orange, banana</ld-accordion-panel>
            </ld-accordion-section>
            <ld-accordion-section>
              <ld-accordion-toggle>Vegetables</ld-accordion-toggle>
              <ld-accordion-panel>Potato, cucumber, tomato</ld-accordion-panel>
            </ld-accordion-section>
            <ld-accordion-section>
              <ld-accordion-toggle disabled>Nuts</ld-accordion-toggle>
              <ld-accordion-panel>
                Walnut, chestnut, strawberry
              </ld-accordion-panel>
            </ld-accordion-section>
          </ld-accordion>
        ),
      })

      const ldAccordion = page.root

      jest.advanceTimersByTime(0)
      await page.waitForChanges()
      await transitionEnd(
        page,
        ldAccordion.querySelectorAll('ld-accordion-panel')[1]
      )

      const handleLdaccordionchange = jest.fn()
      const handleLdaccordiontoggleclick = jest.fn()
      ldAccordion.addEventListener('ldaccordionchange', handleLdaccordionchange)
      ldAccordion.addEventListener(
        'ldaccordiontoggleclick',
        handleLdaccordiontoggleclick
      )

      const ldAccordionToggles = ldAccordion.querySelectorAll(
        'ld-accordion-toggle'
      )
      const triggerBtn0 =
        ldAccordionToggles[0].shadowRoot.querySelector('button')
      const triggerBtn2 =
        ldAccordionToggles[2].shadowRoot.querySelector('button')

      triggerBtn0.dispatchEvent(new Event('click'))

      expect(handleLdaccordionchange).toHaveBeenCalledTimes(1)
      expect(handleLdaccordiontoggleclick).toHaveBeenCalledTimes(1)

      triggerBtn0.dispatchEvent(new Event('click'))
      expect(handleLdaccordionchange).toHaveBeenCalledTimes(2)
      expect(handleLdaccordiontoggleclick).toHaveBeenCalledTimes(2)

      triggerBtn2.dispatchEvent(new Event('click'))
      expect(handleLdaccordionchange).toHaveBeenCalledTimes(2)
      expect(handleLdaccordiontoggleclick).toHaveBeenCalledTimes(2)
    })

    it('emits ldaccordionchange, ldaccordionlabelclick and ldaccordiontoggleclick events in split mode', async () => {
      const page = await newSpecPage({
        components,
        template: () => (
          <ld-accordion>
            <ld-accordion-section expanded>
              <ld-accordion-toggle split>Fruits</ld-accordion-toggle>
              <ld-accordion-panel>Apple, orange, banana</ld-accordion-panel>
            </ld-accordion-section>
            <ld-accordion-section>
              <ld-accordion-toggle split>Vegetables</ld-accordion-toggle>
              <ld-accordion-panel>Potato, cucumber, tomato</ld-accordion-panel>
            </ld-accordion-section>
            <ld-accordion-section>
              <ld-accordion-toggle split disabled>
                Nuts
              </ld-accordion-toggle>
              <ld-accordion-panel>
                Walnut, chestnut, strawberry
              </ld-accordion-panel>
            </ld-accordion-section>
          </ld-accordion>
        ),
      })

      const ldAccordion = page.root

      const handleLdaccordionchange = jest.fn()
      const handleLdaccordionlabelclick = jest.fn()
      const handleLdaccordiontoggleclick = jest.fn()
      ldAccordion.addEventListener('ldaccordionchange', handleLdaccordionchange)
      ldAccordion.addEventListener(
        'ldaccordionlabelclick',
        handleLdaccordionlabelclick
      )
      ldAccordion.addEventListener(
        'ldaccordiontoggleclick',
        handleLdaccordiontoggleclick
      )

      const ldAccordionToggles = ldAccordion.querySelectorAll(
        'ld-accordion-toggle'
      )
      const toggleTriggerBtn0 = ldAccordionToggles[0].shadowRoot.querySelector(
        'button.ld-accordion-toggle__trigger'
      )
      const toggleLabelBtn0 = ldAccordionToggles[0].shadowRoot.querySelector(
        'button.ld-accordion-toggle__label'
      )
      const toggleTriggerBtn2 = ldAccordionToggles[2].shadowRoot.querySelector(
        'button.ld-accordion-toggle__trigger'
      )
      const toggleLabelBtn2 = ldAccordionToggles[2].shadowRoot.querySelector(
        'button.ld-accordion-toggle__label'
      )

      toggleLabelBtn0.dispatchEvent(new Event('click'))
      expect(handleLdaccordionlabelclick).toHaveBeenCalledTimes(1)
      expect(handleLdaccordionchange).toHaveBeenCalledTimes(0)
      expect(handleLdaccordiontoggleclick).toHaveBeenCalledTimes(0)

      toggleTriggerBtn0.dispatchEvent(new Event('click'))
      expect(handleLdaccordionlabelclick).toHaveBeenCalledTimes(1)
      expect(handleLdaccordionchange).toHaveBeenCalledTimes(1)
      expect(handleLdaccordiontoggleclick).toHaveBeenCalledTimes(1)

      toggleLabelBtn2.dispatchEvent(new Event('click'))
      expect(handleLdaccordionlabelclick).toHaveBeenCalledTimes(1)
      expect(handleLdaccordionchange).toHaveBeenCalledTimes(1)
      expect(handleLdaccordiontoggleclick).toHaveBeenCalledTimes(1)

      toggleTriggerBtn2.dispatchEvent(new Event('click'))
      expect(handleLdaccordionlabelclick).toHaveBeenCalledTimes(1)
      expect(handleLdaccordionchange).toHaveBeenCalledTimes(1)
      expect(handleLdaccordiontoggleclick).toHaveBeenCalledTimes(1)
    })
  })

  it('does not throw when disconnecting before hydration', () => {
    const component = new LdAccordionPanel()
    component.disconnectedCallback()
  })
})
