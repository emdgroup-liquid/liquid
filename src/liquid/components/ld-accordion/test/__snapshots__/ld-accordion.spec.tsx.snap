// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-accordion renders 1`] = `
<ld-accordion class="ld-accordion">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <ld-accordion-section class="ld-accordion-section ld-accordion-section--expanded">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    <ld-accordion-toggle class="ld-accordion-toggle ld-accordion-toggle--expanded">
      <mock:shadow-root>
        <button aria-expanded="true" class="ld-accordion-toggle__button" part="toggle focusable">
          <div class="ld-accordion-toggle__content" part="content">
            <div class="ld-accordion-toggle__label" part="label">
              <div class="ld-accordion-toggle__label-content" part="label-content">
                <slot></slot>
              </div>
            </div>
            <div class="ld-accordion-toggle__trigger" part="trigger">
              <div class="ld-accordion-toggle__trigger-content" part="trigger-content">
                <slot name="icon"></slot>
                <ld-icon aria-hidden="true" class="ld-accordion-toggle__trigger-icon" name="arrow-down" part="trigger-icon" size="sm"></ld-icon>
              </div>
            </div>
          </div>
        </button>
      </mock:shadow-root>
      Fruits
    </ld-accordion-toggle>
    <ld-accordion-panel class="ld-accordion-panel ld-accordion-panel--expanded" style="--ld-accordion-panel-max-height: NaNpx;">
      <mock:shadow-root>
        <div class="ld-accordion-panel__content" part="content">
          <slot></slot>
        </div>
      </mock:shadow-root>
      Apple, orange, banana
    </ld-accordion-panel>
  </ld-accordion-section>
  <ld-accordion-section class="ld-accordion-section">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    <ld-accordion-toggle class="ld-accordion-toggle">
      <mock:shadow-root>
        <button aria-expanded="false" class="ld-accordion-toggle__button" part="toggle focusable">
          <div class="ld-accordion-toggle__content" part="content">
            <div class="ld-accordion-toggle__label" part="label">
              <div class="ld-accordion-toggle__label-content" part="label-content">
                <slot></slot>
              </div>
            </div>
            <div class="ld-accordion-toggle__trigger" part="trigger">
              <div class="ld-accordion-toggle__trigger-content" part="trigger-content">
                <slot name="icon"></slot>
                <ld-icon aria-hidden="true" class="ld-accordion-toggle__trigger-icon" name="arrow-down" part="trigger-icon" size="sm"></ld-icon>
              </div>
            </div>
          </div>
        </button>
      </mock:shadow-root>
      Vegetables
    </ld-accordion-toggle>
    <ld-accordion-panel class="ld-accordion-panel" style="--ld-accordion-panel-max-height: 0px;">
      <mock:shadow-root>
        <div class="ld-accordion-panel__content" part="content">
          <slot></slot>
        </div>
      </mock:shadow-root>
      Potato, cucumber, tomato
    </ld-accordion-panel>
  </ld-accordion-section>
  <ld-accordion-section class="ld-accordion-section">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    <ld-accordion-toggle class="ld-accordion-toggle">
      <mock:shadow-root>
        <button aria-disabled="true" aria-expanded="false" class="ld-accordion-toggle__button" part="toggle focusable">
          <div class="ld-accordion-toggle__content" part="content">
            <div class="ld-accordion-toggle__label" part="label">
              <div class="ld-accordion-toggle__label-content" part="label-content">
                <slot></slot>
              </div>
            </div>
            <div class="ld-accordion-toggle__trigger" part="trigger">
              <div class="ld-accordion-toggle__trigger-content" part="trigger-content">
                <slot name="icon"></slot>
                <ld-icon aria-hidden="true" class="ld-accordion-toggle__trigger-icon" name="arrow-down" part="trigger-icon" size="sm"></ld-icon>
              </div>
            </div>
          </div>
        </button>
      </mock:shadow-root>
      Nuts
    </ld-accordion-toggle>
    <ld-accordion-panel class="ld-accordion-panel" style="--ld-accordion-panel-max-height: 0px;">
      <mock:shadow-root>
        <div class="ld-accordion-panel__content" part="content">
          <slot></slot>
        </div>
      </mock:shadow-root>
      Walnut, chestnut, strawberry
    </ld-accordion-panel>
  </ld-accordion-section>
</ld-accordion>
`;
